app/services/badge_service.rb

```diff
@@ -0,0 +1,47 @@
+class BadgeService
+  def initialize(test_passage)
+    @test_passage = test_passage
+    @user = test_passage.user
+    @test = test_passage.test
+  end
+
+  def badges
+    Badge.all.select do |badge|
+      badge if send(badge.rule_name, badge)
+    end
+  end
+
+  private
+
+  def all_test_from_category(badge)
+    return false unless badge.category == @test.category
+
+    if @user.badges.where('users_badges.badge_id = ?', badge.id).length != 0
```

Comment:
```suggestion
    if @user.badges.where(id: badge.id).exists?
```
к соединительной таблице можно не обращаться, просто смотришь есть уже такие бейджи у пользователя или нет, если один и тот же бейдж может выдаваться многократно, то значит у пользователя может быть несколько бейджей с одинаковым ай-ди

про `exists?`: https://guides.rubyonrails.org/active_record_querying.html#existence-of-objects